packages/downloading-helpers/src/scripts/replay-assets.ts

```diff
@@ -58,6 +65,7 @@ export const fetchAsset: (path: string) => Promise<string> = async (path) => {
     }
     await saveAssetMetadata(assetMetadata);
     releaseLock();
+    await fetchSourceMapIfDebugging(path);
```

Comment:
[nit] I found the recursion made it a bit slower for me to parse what's going on. For fast readability you could also rename fetchAsset to fetchJsFile, have fetchAsset just download an arbitrary asset form a URL, and then communicate the intent more directly by doing:

```ts
/**
 * Downloads the given 'module' JS file as an MJS file (to force correct interpretation when loading the JS file).
 *
 * The associated source map will also be downloaded if present to sit alongside the main JS file and if the base snippets URL is a localhost URL.  
 */
export const fetchJsFile = (path: string): Promise<string> => {
     const snippetsBaseUrl = getSnippetsBaseUrl();
     const fetchUrl = new URL(path, getSnippetsBaseUrl()).href;
     const assetFileName = basename(new URL(fetchUrl).pathname);
     
     const jsFilePath = await fetchAsset(fetchUrl, convertJsExtensionToCjs(assetFileName));
     if (snippetsBaseUrl.includes(localhost)) {
           await fetchAsset(`${fetchUrl}.map`, `${assetFileName}.map`);
     }
     
     return jsFilePath;
}

const fetchAsset = (urlToDownloadFrom: string, fileNameToDownloadAs: string): Promise<string> => {
  // ... original version of fetch asset but without .js specific code: just generic utility for downloading & caching assets ...
}
```

fetchAsset would then be private to the file / not exported

This avoids the reader having to follow the recursion in their mind, convince themselves that it only goes once deep etc. (as soon as they see the recursion they'll probably have questions around at what point does it terminate). The code above communicates the intent and pattern more directly: download the .js file, and then maybe download the .map file.

It also has the slight advantage of colocating the stuff around snippetsBaseUrl / code that changes together lives together. And the seperation is clean: unless they are interested in the caching, most people interested in the code probably won't need to read fetchAsset at all, just fetchJsFile to see what it does.

Side note: fetchAsset is used in the main Meticulous repo as well, but that's OK: we can just switch to the new method name next time we bump the main repo.

Reply:
Also, wonder if we should make it not fail if the .map file doesn't exist?

Reply:
This is indeed much nicer and avoids the recursion, done. I went for `fetchAsset` and `fetchAndCacheFile` so that we avoid the breaking change in the name. 

> Also, wonder if we should make it not fail if the .map file doesn't exist?

I'd prefer to not. Obviously it means a slight annoyance that we will fail if the snippet server is an older version that isn't serving sourcemaps. But I think keeping things simple and making sure stuff breaks loudly if in the future if we're missing source maps (e.g. we change our build and it stops producing them in the right place) is better.